### PR TITLE
[DataLayer] Implement listener selector

### PR DESCRIPTION
**Pull request by bpauli**
- match listener only for given selector if provided
- add unit tests

### CODE REVIEW

File: src/scripts/DataLayer.js [hunk 1]
```diff
@@ -313,21 +313,41 @@ DataLayer.Manager.prototype._isMatching = function(listener, item) {
...
       listenerConfig.on === DataLayer.constants.event.CHANGE) {
-      isMatching = true;
+      isMatching = this._isSelectorMatching(listenerConfig, itemConfig);
     }
   }
   return isMatching;
 };
 
+/**
+ * Checks if the given listerConfig has a selector and if so, is it matching with the data payload of the itemConfig.
```
Comment (jckautzmann) [inline src/scripts/DataLayer.js]: I'd rephrase it e.g.:
Checks if the given listenerConfig has a selector that points to an object in the data payload of the itemConfig.

File: src/scripts/DataLayer.js [hunk 2]
```diff
@@ -313,21 +313,41 @@ DataLayer.Manager.prototype._isMatching = function(listener, item) {
...
+ *
+ * @param {DataLayer.Item} listenerConfig Config of the listener to extract the selector from.
+ * @param {DataLayer.Item} itemConfig The item config withe data.
```
Comment (jckautzmann) [inline src/scripts/DataLayer.js]: the type should be `Object`

File: src/scripts/DataLayer.js [hunk 3]
```diff
@@ -313,21 +313,41 @@ DataLayer.Manager.prototype._isMatching = function(listener, item) {
...
+ * @returns {Boolean} true if no selector is provided or the given selector is matching, false otherwise.
```
Comment (jckautzmann) [inline src/scripts/DataLayer.js]: I'd rephrase it e.g.:
true if a selector is not provided or if the given selector is matching, false otherwise.

File: src/scripts/DataLayerUtils.js [hunk 4]
```diff
@@ -48,6 +48,33 @@ const utils = {
...
+  },
+
+  /**
+   * Checks if the given object has the nested property.
+   *
+   * @param {String} propertyPath The property to be be checked
+   * @param {Object} obj The object that will be checked.
+   * @returns {Boolean} true if the given object has the nested property, false otherwise.
+   */
+  hasOwnNestedProperty: function(propertyPath, obj) {
```
Comment (jckautzmann) [inline src/scripts/DataLayerUtils.js]: we may want to use the lodash `has` function instead of having our own implementation:
https://lodash.com/docs#has

File: src/tests/DataLayer.test.js [hunk 5]
```diff
@@ -362,6 +362,101 @@ test('listener on: register a handler (with a static function) that has already
...
+  test('on change listener with selector for image component data', () => {
+    dataLayer.push(argOn);
+    dataLayer.push({ data: carouselData });
+    expect(mockCallback.mock.calls.length).toBe(0);
+    dataLayer.push({ data: imageData });
+    expect(mockCallback.mock.calls.length).toBe(1);
+  });
+
+  test('custom listener with selector for image component data', () => {
+    let argOn = {
```
Comment (jckautzmann) [inline src/tests/DataLayer.test.js]: I'd call it `listenerOn`

**Comment by jckautzmann** [comment]
Should we also change the logic of the handler to pass the old and the new value of the matching object?
`handler: function(item, oldValue, newValue)`

**Comment by richardhand** [comment]
There are some merge conflicts, since we recently replaced the utils with `lodash.isEqual` and `lodash.merge`.
